### PR TITLE
feat: improve checker functionality when `Aggregate` is used as a parameter

### LINKED ISSUE #509: No warning raised when a function is used in a loop with `Aggregate` as parameter instead as extension receiver
The following code builds correctly with no issue but a warning should be raised

```kt
private fun Aggregate<Int>.foo() {
    for (i in 1..10) {
        bar(this, i)
    }
}

fun bar(aggregate: Aggregate<Int>, i: Int) = aggregate.neighboring(i)
```

@FreshMag could you put a regression test in the codebase and fix it with a PR?
If additional corner cases are spotted out, insert them as a regression test

### PR DESCRIPTION
**Pull request by FreshMag**
closes #509 

The checker for Aggregate usages inside iterations now also works when Aggregate is used as a parameter instead of as an extension receiver.

As you can see from this PR, I’ve frequently used the term “delegate” to describe a function that takes an Aggregate parameter and uses it.

The need to check for Aggregate parameters in delegate functions introduces several edge cases, which I’ve tried to encapsulate in the testing resources as comprehensively as possible. Side note: some of these tests have very long descriptions, which might benefit from being made shorter.

Currently, three cases remain uncaptured by the checker. More accurately, there is one primary case, while the other two are “variants” involving the presence of alignedOn. The main case is as follows:

```kotlin
fun delegate(aggregate: Aggregate<Int>) {
  fun delegate2() {
    aggregate.neighboring(0)
  }
  delegate2()
}

fun Aggregate<Int>.entry() {
  for (i in 1..3) {
    delegate(this)
  }
}
```
This scenario is not yet detected by the checker. Please let me know if I should prioritize addressing this now or if it can be considered an edge case to defer for the time being.

### CODE REVIEW

**Comment by DanySK** [comment]
generated code must not be tracked

**Comment by DanySK** [comment]
@FreshMag
uuhmm... How general is your tool?
If it is generally applicable, it should be converted to a Gradle plugin of yours.
If it is collektive-specific, then it can be added to the plugins that we have in this project

Where can I take a look to the tool?

**Comment by FreshMag** [comment]
> @FreshMag uuhmm... How general is your tool? If it is generally applicable, it should be converted to a Gradle plugin of yours. If it is collektive-specific, then it can be added to the plugins that we have in this project

I started it as a general tool to generate tests based on some rules or configurations. I was thinking of publishing it on Maven eventually. For now it's a very small tool that simply works using template engines.

> Where can I take a look to the tool?

[Here's the link to the repository](https://github.com/FreshMag/subjekt). Please note that the code is not perfect, lacks documentation, and contains many hard-coded or temporary elements.

**Comment by DanySK** [comment]
@FreshMag that thing is super nice, you should publish it on Central ASAP, build a gradle plugin out of it, and then use it to generate tests. Then you can improve it at will.

And, of course, it must be part of your thesis dissertation, as we learned that tests cases can be generated.

**Comment by FreshMag** [comment]
@DanySK thank you. In the next days I will clean the code and publish it.

**Comment by nicolasfara** [comment]
> > @FreshMag uuhmm... How general is your tool? If it is generally applicable, it should be converted to a Gradle plugin of yours. If it is collektive-specific, then it can be added to the plugins that we have in this project
> 
> I started it as a general tool to generate tests based on some rules or configurations. I was thinking of publishing it on Maven eventually. For now it's a very small tool that simply works using template engines.
> 
> > Where can I take a look to the tool?
> 
> [Here's the link to the repository](https://github.com/FreshMag/subjekt). Please note that the code is not perfect, lacks documentation, and contains many hard-coded or temporary elements.

@FreshMag That's amazing 😍 good job! How have you generated the logo? It is gorgeous!! 😄

**Comment by FreshMag** [comment]
> @FreshMag That's amazing 😍 good job! How have you generated the logo? It is gorgeous!! 😄

Thank you! I made the logo using Figma.

**Comment by FreshMag** [comment]
> In the next days I will clean the code and publish it.

I apologize for the delay; it took me a while. I’ve deleted the generated files. The project is currently on Maven with the ID `subjekt-api`. Additionally, I’ve published a simple Gradle plugin to simplify library usage. However, there appears to be an issue with including the `jackson` library in the plugin when attempting to integrate it with Collektive, likely caused by a version conflict. I’ll work on resolving this soon. In the meantime, I’m using Subjekt directly within Collektive’s tests.

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -59,4 +60,5 @@ kotlin-qa = "org.danilopianini.gradle-kotlin-qa:0.78.0"
 kover = "org.jetbrains.kotlinx.kover:0.8.3"
 multiJvmTesting = "org.danilopianini.multi-jvm-test-plugin:3.0.0"
 publishOnCentral = "org.danilopianini.publish-on-central:7.0.0"
+subjekt = "io.github.freshmag.subjekt:1.0.1"
```
Comment (DanySK) [inline gradle/libs.versions.toml:63]: If it is not used, maybe let's delay its inclusion

**Review by DanySK — APPROVED** [review]
@nicolasfara, I leave the details to you. Conceptually seems good to me

File: compiler-plugin-test/src/test/resources/subjekt/IterationWithAggregate.yaml [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+          Consider to wrap the function into the 'alignedOn' method with a unique element.
+
+  - name: |-
+      IterationExtAlign${{ prettify(AGGREGATE, loop(AGGREGATE)) }}
+    code: |-
+      fun Aggregate<Int>.entry() {
+        ${{ alignedOn(loop(AGGREGATE)) }}
+      }
+    outcomes:
+      - warning: |-
```
Comment (DanySK) [inline compiler-plugin-test/src/test/resources/subjekt/IterationWithAggregate.yaml:67]: Suggested improvement for Subjective: support direct elements with no list symbol (Konf does it automatically)

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/IterationUsingDelegatesWithoutAlignSpec.kt [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                 * Gets the text from a map of files, given its [caseName], and converts it to a
+                 * [KotlinTestingProgram].
+                 */
+                fun getTestingProgram(caseName: String): KotlinTestingProgram =
+                    testSubjects[pascalCase(caseName, functionName, iteration)]
+                        ?.readText()
+                        ?.asTestingProgram("$functionName-${caseName}_$iteration.kt")
+                        ?: throw FileNotFoundException(
+                            "Program not found: ${pascalCase(caseName, functionName, iteration)}",
+                        )
```
Comment (nicolasfara) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/IterationUsingDelegatesWithoutAlignSpec.kt:56]:
```suggestion
                fun getTestingProgram(caseName: String): KotlinTestingProgram =
                    val subject = pascalCase(caseName, functionName, iteration)
                    testSubjects[subject]
                        ?.readText()
                        ?.asTestingProgram("$functionName-${caseName}_$iteration.kt")
                        ?: throw FileNotFoundException(
                            "Program not found: $subject",
                        )
```

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/IterationUsingDelegatesWithoutAlignSpec.kt [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+//
+//                "inside $iterationDescription and using a function that takes an Aggregate argument, " +
+//                        "using it with a specific alignedOn, inside a called nested function" - {
+//                    val case = "IterationDelegatedNestedFunAlign"
+//                    val code = getTestingProgram(case)
+//
+//                    "should compile producing a warning" - {
+//                        code shouldCompileWith noWarning
+//                    }
+//                }
```
Comment (nicolasfara) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/IterationUsingDelegatesWithoutAlignSpec.kt:169]: Instead of commenting out these test cases, you could disable them with `.config(enabled = false)`.
This way, we have the tests in the codebase but temporarily disabled.

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/IterationWithoutAlignSpec.kt [hunk 5]
```diff
@@ -24,22 +25,28 @@ class IterationWithoutAlignSpec : FreeSpec({
...
-                        case = caseName,
-                        iteration = iteration,
-                        aggregateFunction = functionName,
-                    ).asTestingProgram("$functionName-${caseName}_$iteration.kt")
+                    testSubjects[pascalCase(caseName, functionName, iteration)]
+                        ?.readText()
+                        ?.asTestingProgram("$functionName-${caseName}_$iteration.kt")
+                        ?: throw FileNotFoundException(
+                            "Program not found: ${pascalCase(caseName, functionName, iteration)}",
+                        )
```
Comment (nicolasfara) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/IterationWithoutAlignSpec.kt:46]:
```suggestion
                    val subject = pascalCase(caseName, functionName, iteration)
                    testSubjects[subject]
                        ?.readText()
                        ?.asTestingProgram("$functionName-${caseName}_$iteration.kt")
                        ?: throw FileNotFoundException(
                            "Program not found: $subject",
                        )
```

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/CheckersUtility.kt [hunk 6]
```diff
@@ -153,4 +159,24 @@ object CheckersUtility {
...
+                parameter.resolvedReturnTypeRef.coneType
+            }
+
+    /**
+     * Checks whether if the called function accepts at least on argument of type
+     * [it.unibo.collektive.aggregate.api.Aggregate].
+     */
+    fun FirFunctionCall.hasAggregateArgument(): Boolean =
+        getArgumentsTypes()?.any {
+            it.classId == ClassId.fromString(AGGREGATE_CLASS_FQ_NAME.replace(".", "/"))
```
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/CheckersUtility.kt:180]: Can we do it better? Does `ClassId.fromString` not accept FQN classes?

What about:

```kotlin
ClassId.topLevel(FqName(AGGREGATE_CLASS_FQ_NAME))
```

Could you verify if it works? Additionally, we can introduce a `fun String.toFqNameUnsafe(): FqName` to easily convert a `String` into a `FqName`. What do you think?

**Review by nicolasfara — CHANGES_REQUESTED** [review]
Reviewed 19/21

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/FunctionCallWithAggregateParVisitor.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    override fun visitFunctionCall(functionCall: FirFunctionCall) {
+        if (functionCall.isAggregate(context.session)) {
+            if (functionCall.fqName() == ALIGNED_ON_FUNCTION_FQ_NAME) {
+                insideAlignedOn = true
+                functionCall.acceptChildren(this)
+                insideAlignedOn = false
+            } else if (!isInsideAlignedOnOrNestedFun()) {
+                found = true
+                return
```
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/FunctionCallWithAggregateParVisitor.kt:48]: I think those explicit `return`s can be avoided.

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/FunctionCallWithAggregateParVisitor.kt [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            found = visitor.visitSuspiciousFunctionCallDeclaration(functionCall)
+        }
+        return
+    }
+
+    override fun visitSimpleFunction(simpleFunction: FirSimpleFunction) {
+        functionCounter++
+        simpleFunction.body?.accept(this)
+        functionCounter--
+        return
```
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/FunctionCallWithAggregateParVisitor.kt:61]: Ditto

**Review by nicolasfara — CHANGES_REQUESTED** [review]
